File: backend/app/schemas/resume.py
"""
简历相关的 Pydantic 模式
"""

from datetime import datetime, date
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict

# ============ 简历内容子模块 ============


class BasicInfo(BaseModel):
    """基本信息"""

    name: str = Field(..., description="姓名")
    gender: Optional[str] = Field(None, description="性别")
    birth_date: Optional[date] = Field(None, description="出生日期")
    phone: Optional[str] = Field(None, description="手机号")
    email: Optional[str] = Field(None, description="邮箱")
    location: Optional[str] = Field(None, description="所在城市")
    avatar_url: Optional[str] = Field(None, description="头像URL")
    job_intention: Optional[str] = Field(None, description="求职意向")
    expected_salary: Optional[str] = Field(None, description="期望薪资")
    self_introduction: Optional[str] = Field(None, description="个人简介")


class Education(BaseModel):
    """教育经历"""

    school: str = Field(..., description="学校名称")
    major: Optional[str] = Field(None, description="专业")
    degree: Optional[str] = Field(None, description="学历")
    start_date: Optional[date] = Field(None, description="开始时间")
    end_date: Optional[date] = Field(None, description="结束时间")
    gpa: Optional[str] = Field(None, description="GPA")
    description: Optional[str] = Field(None, description="描述")


class WorkExperience(BaseModel):
    """工作经历"""

    company: str = Field(..., description="公司名称")
    position: str = Field(..., description="职位")
    department: Optional[str] = Field(None, description="部门")
    start_date: Optional[date] = Field(None, description="开始时间")
    end_date: Optional[date] = Field(None, description="结束时间")
    is_current: bool = Field(False, description="是否在职")
    description: Optional[str] = Field(None, description="工作描述")
    achievements: Optional[List[str]] = Field(None, description="工作成就")


class Project(BaseModel):
    """项目经历"""

    name: str = Field(..., description="项目名称")
    role: Optional[str] = Field(None, description="担任角色")
    start_date: Optional[date] = Field(None, description="开始时间")
    end_date: Optional[date] = Field(None, description="结束时间")
    description: Optional[str] = Field(None, description="项目描述")
    tech_stack: Optional[List[str]] = Field(None, description="技术栈")
    achievements: Optional[List[str]] = Field(None, description="项目成果")
    link: Optional[str] = Field(None, description="项目链接")


class Skill(BaseModel):
    """技能"""

    name: str = Field(..., description="技能名称")
    level: Optional[str] = Field(None, description="掌握程度")
    category: Optional[str] = Field(None, description="技能分类")


class Certification(BaseModel):
    """证书"""

    name: str = Field(..., description="证书名称")
    issuer: Optional[str] = Field(None, description="颁发机构")
    issue_date: Optional[date] = Field(None, description="获得时间")
    expiry_date: Optional[date] = Field(None, description="有效期")
    credential_id: Optional[str] = Field(None, description="证书编号")


class CustomSection(BaseModel):
    """自定义模块"""

    title: str = Field(..., description="模块标题")
    content: str = Field(..., description="模块内容")
    order: int = Field(0, description="排序")


# ============ 完整简历内容 ============


class ResumeContent(BaseModel):
    """简历结构化内容"""

    basic_info: Optional[BasicInfo] = None
    education: Optional[List[Education]] = Field(default_factory=list)
    work_experience: Optional[List[WorkExperience]] = Field(default_factory=list)
    projects: Optional[List[Project]] = Field(default_factory=list)
    skills: Optional[List[Skill]] = Field(default_factory=list)
    certifications: Optional[List[Certification]] = Field(default_factory=list)
    custom_sections: Optional[List[CustomSection]] = Field(default_factory=list)


class StyleConfig(BaseModel):
    """样式配置"""

    theme: str = Field("default", description="主题")
    primary_color: str = Field("#2B2B2B", description="主色调")
    secondary_color: str = Field("#666666", description="辅助色")
    font_family: str = Field("Microsoft YaHei", description="字体")
    font_size: int = Field(12, description="字号")
    line_height: float = Field(1.5, description="行高")
    margin: Dict[str, int] = Field(
        default_factory=lambda: {"top": 20, "right": 20, "bottom": 20, "left": 20}
    )
    layout: str = Field("single", description="布局: single/double")


# ============ API 请求/响应模式 ============


class ResumeCreate(BaseModel):
    """创建简历请求"""

    title: str = Field(..., min_length=1, max_length=255, description="简历标题")
    description: Optional[str] = None
    template_id: Optional[int] = None
    content: Optional[ResumeContent] = None
    style_config: Optional[StyleConfig] = None


class ResumeUpdate(BaseModel):
    """更新简历请求"""

    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    template_id: Optional[int] = None
    content: Optional[ResumeContent] = None
    style_config: Optional[StyleConfig] = None
    status: Optional[str] = None


class ResumeResponse(BaseModel):
    """简历响应"""

    id: int
    user_id: int
    title: str
    description: Optional[str] = None
    template_id: Optional[int] = None
    content: Optional[Dict[str, Any]] = None
    style_config: Optional[Dict[str, Any]] = None
    status: str
    version: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeListResponse(BaseModel):
    """简历列表响应"""

    items: List[ResumeResponse]
    total: int
    page: int
    page_size: int


class ResumeVersionResponse(BaseModel):
    """简历版本响应"""

    id: int
    resume_id: int
    version_number: int
    content: Dict[str, Any]
    style_config: Optional[Dict[str, Any]] = None
    change_note: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============ AI 生成相关 ============


class AIGenerateRequest(BaseModel):
    """AI生成简历请求"""

    target_position: str = Field(..., description="目标岗位")
    user_background: Optional[str] = Field(None, description="用户背景描述")
    style: str = Field("professional", description="风格: professional/creative/academic")
    language: str = Field("zh", description="语言: zh/en")


class AIOptimizeRequest(BaseModel):
    """AI优化内容请求"""

    content: str = Field(..., description="需要优化的内容")
    optimization_type: str = Field(
        "star_method", description="优化类型: star_method/quantify/keywords/polish"
    )
    context: Optional[str] = Field(None, description="上下文信息")


class AIOptimizeResponse(BaseModel):
    """AI优化响应"""

    original: str
    optimized: str
    suggestions: Optional[List[str]] = None

File: ai-resume-flutter/lib/data/models/resume.dart
import 'package:json_annotation/json_annotation.dart';

part 'resume.g.dart';

/// 简历数据模型
@JsonSerializable()
class Resume {
  final int id;
  final String title;
  final ResumeContent content;
  final int status; // 0: 草稿, 1: 已发布, 2: 已归档
  final DateTime createdAt;
  final DateTime updatedAt;

  Resume({
    required this.id,
    required this.title,
    required this.content,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Resume.fromJson(Map<String, dynamic> json) =>
      _$ResumeFromJson(json);

  Map<String, dynamic> toJson() => _$ResumeToJson(this);

  Resume copyWith({
    int? id,
    String? title,
    ResumeContent? content,
    int? status,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Resume(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// 简历内容
@JsonSerializable()
class ResumeContent {
  final BasicInfo? basicInfo;
  final List<Education>? education;
  final List<WorkExperience>? workExperience;
  final List<Project>? projects;
  final List<Skill>? skills;

  ResumeContent({
    this.basicInfo,
    this.education,
    this.workExperience,
    this.projects,
    this.skills,
  });

  factory ResumeContent.fromJson(Map<String, dynamic> json) =>
      _$ResumeContentFromJson(json);

  Map<String, dynamic> toJson() => _$ResumeContentToJson(this);

  ResumeContent copyWith({
    BasicInfo? basicInfo,
    List<Education>? education,
    List<WorkExperience>? workExperience,
    List<Project>? projects,
    List<Skill>? skills,
  }) {
    return ResumeContent(
      basicInfo: this.basicInfo ?? basicInfo,
      education: this.education ?? education,
      workExperience: this.workExperience ?? workExperience,
      projects: this.projects ?? projects,
      skills: this.skills ?? skills,
    );
  }

  /// 创建空内容
  factory ResumeContent.empty() {
    return ResumeContent(
      basicInfo: BasicInfo.empty(),
      education: [],
      workExperience: [],
      projects: [],
      skills: [],
    );
  }
}

/// 基本信息
@JsonSerializable()
class BasicInfo {
  final String? name;
  final String? email;
  final String? phone;
  final String? location;
  final String? title;
  final String? summary;
  final String? jobIntention;
  final String? selfIntroduction;
  final String? avatar;
  final String? linkedin;
  final String? github;
  final String? website;

  BasicInfo({
    this.name,
    this.email,
    this.phone,
    this.location,
    this.title,
    this.summary,
    this.jobIntention,
    this.selfIntroduction,
    this.avatar,
    this.linkedin,
    this.github,
    this.website,
  });

  factory BasicInfo.fromJson(Map<String, dynamic> json) =>
      _$BasicInfoFromJson(json);

  Map<String, dynamic> toJson() => _$BasicInfoToJson(this);

  BasicInfo copyWith({
    String? name,
    String? email,
    String? phone,
    String? location,
    String? title,
    String? summary,
    String? jobIntention,
    String? selfIntroduction,
    String? avatar,
    String? linkedin,
    String? github,
    String? website,
  }) {
    return BasicInfo(
      name: name ?? this.name,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      location: location ?? this.location,
      title: title ?? this.title,
      summary: summary ?? this.summary,
      jobIntention: jobIntention ?? this.jobIntention,
      selfIntroduction: selfIntroduction ?? this.selfIntroduction,
      avatar: avatar ?? this.avatar,
      linkedin: linkedin ?? this.linkedin,
      github: github ?? this.github,
      website: website ?? this.website,
    );
  }

  /// 创建空基本信息
  factory BasicInfo.empty() {
    return BasicInfo();
  }

  /// 是否为空
  bool get isEmpty {
    return name == null &&
        email == null &&
        phone == null &&
        location == null &&
        title == null &&
        summary == null &&
        jobIntention == null &&
        selfIntroduction == null;
  }
}

/// 教育经历
@JsonSerializable()
class Education {
  final String school;
  final String degree;
  final String? major;
  final String? startDate;
  final String? endDate;
  final String? gpa;
  final String? description;

  Education({
    required this.school,
    required this.degree,
    this.major,
    this.startDate,
    this.endDate,
    this.gpa,
    this.description,
  });

  factory Education.fromJson(Map<String, dynamic> json) =>
      _$EducationFromJson(json);

  Map<String, dynamic> toJson() => _$EducationToJson(this);

  Education copyWith({
    String? school,
    String? degree,
    String? major,
    String? startDate,
    String? endDate,
    String? gpa,
    String? description,
  }) {
    return Education(
      school: school ?? this.school,
      degree: degree ?? this.degree,
      major: major ?? this.major,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      gpa: gpa ?? this.gpa,
      description: description ?? this.description,
    );
  }

  /// 创建空教育经历
  factory Education.empty() {
    return Education(
      school: '',
      degree: '',
    );
  }
}

/// 工作经历
@JsonSerializable()
class WorkExperience {
  final String company;
  final String position;
  final String? startDate;
  final String? endDate;
  final bool? isCurrent;
  final String? location;
  final String? description;
  final List<String>? achievements;

  WorkExperience({
    required this.company,
    required this.position,
    this.startDate,
    this.endDate,
    this.isCurrent,
    this.location,
    this.description,
    this.achievements,
  });

  factory WorkExperience.fromJson(Map<String, dynamic> json) =>
      _$WorkExperienceFromJson(json);

  Map<String, dynamic> toJson() => _$WorkExperienceToJson(this);

  WorkExperience copyWith({
    String? company,
    String? position,
    String? startDate,
    String? endDate,
    bool? isCurrent,
    String? location,
    String? description,
    List<String>? achievements,
  }) {
    return WorkExperience(
      company: company ?? this.company,
      position: position ?? this.position,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      isCurrent: isCurrent ?? this.isCurrent,
      location: location ?? this.location,
      description: description ?? this.description,
      achievements: achievements ?? this.achievements,
    );
  }

  /// 创建空工作经历
  factory WorkExperience.empty() {
    return WorkExperience(
      company: '',
      position: '',
    );
  }
}

/// 项目经历
@JsonSerializable()
class Project {
  final String name;
  final String? role;
  final String? startDate;
  final String? endDate;
  final String? description;
  final List<String>? techStack;
  final List<String>? highlights;
  final String? link;

  Project({
    required this.name,
    this.role,
    this.startDate,
    this.endDate,
    this.description,
    this.techStack,
    this.highlights,
    this.link,
  });

  factory Project.fromJson(Map<String, dynamic> json) =>
      _$ProjectFromJson(json);

  Map<String, dynamic> toJson() => _$ProjectToJson(this);

  Project copyWith({
    String? name,
    String? role,
    String? startDate,
    String? endDate,
    String? description,
    List<String>? techStack,
    List<String>? highlights,
    String? link,
  }) {
    return Project(
      name: name ?? this.name,
      role: role ?? this.role,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      description: description ?? this.description,
      techStack: techStack ?? this.techStack,
      highlights: highlights ?? this.highlights,
      link: link ?? this.link,
    );
  }

  /// 创建空项目
  factory Project.empty() {
    return Project(name: '');
  }
}

/// 技能
@JsonSerializable()
class Skill {
  final String name;
  final String? category;
  final int? level;
  final List<String>? keywords;

  Skill({
    required this.name,
    this.category,
    this.level,
    this.keywords,
  });

  factory Skill.fromJson(Map<String, dynamic> json) =>
      _$SkillFromJson(json);

  Map<String, dynamic> toJson() => _$SkillToJson(this);

  Skill copyWith({
    String? name,
    String? category,
    int? level,
    List<String>? keywords,
  }) {
    return Skill(
      name: name ?? this.name,
      category: category ?? this.category,
      level: level ?? this.level,
      keywords: keywords ?? this.keywords,
    );
  }

  /// 创建空技能
  factory Skill.empty() {
    return Skill(name: '');
  }
}

/// 创建/更新简历请求
@JsonSerializable()
class ResumeRequest {
  final String title;
  final ResumeContent content;

  ResumeRequest({
    required this.title,
    required this.content,
  });

  factory ResumeRequest.fromJson(Map<String, dynamic> json) =>
      _$ResumeRequestFromJson(json);

  Map<String, dynamic> toJson() => _$ResumeRequestToJson(this);
}

/// AI生成简历请求
@JsonSerializable()
class AIGenerateRequest {
  final String targetPosition;

  AIGenerateRequest({
    required this.targetPosition,
  });

  factory AIGenerateRequest.fromJson(Map<String, dynamic> json) =>
      _$AIGenerateRequestFromJson(json);

  Map<String, dynamic> toJson() => _$AIGenerateRequestToJson(this);
}
